src/VirtoCommerce.Platform.Core/Extensions/IQueryableExtensions.cs

```diff
@@ -3,102 +3,230 @@
 using System.Linq;
 using System.Linq.Expressions;
 using System.Reflection;
+using System.Runtime.ExceptionServices;
 
 namespace VirtoCommerce.Platform.Core.Common
 {
     public static class IQueryableExtensions
     {
         public static IOrderedQueryable<T> OrderBy<T>(this IQueryable<T> source, string property)
         {
-            return ApplyOrder<T>(source, property, "OrderBy");
+            return ApplyOrder<T>(source, property, nameof(OrderBy));
         }
+
         public static IOrderedQueryable<T> OrderByDescending<T>(this IQueryable<T> source, string property)
         {
-            return ApplyOrder<T>(source, property, "OrderByDescending");
+            return ApplyOrder<T>(source, property, nameof(OrderByDescending));
         }
+
         public static IOrderedQueryable<T> ThenBy<T>(this IOrderedQueryable<T> source, string property)
         {
-            return ApplyOrder<T>(source, property, "ThenBy");
+            return ApplyOrder<T, IOrderedQueryable<T>>(source, property, nameof(ThenBy), false);
         }
+
         public static IOrderedQueryable<T> ThenByDescending<T>(this IOrderedQueryable<T> source, string property)
         {
-            return ApplyOrder<T>(source, property, "ThenByDescending");
+            return ApplyOrder<T, IOrderedQueryable<T>>(source, property, nameof(ThenByDescending), false);
         }
 
 
         public static IOrderedQueryable<T> OrderBySortInfos<T>(this IQueryable<T> source, IEnumerable<SortInfo> sortInfos)
         {
             if (sortInfos.IsNullOrEmpty())
             {
-                throw new ArgumentNullException("sortInfos");
+                throw new ArgumentNullException(nameof(sortInfos));
             }
-            IOrderedQueryable<T> retVal = null;
-            var firstSortInfo = sortInfos.First();
-            if (firstSortInfo.SortDirection == SortDirection.Descending)
+
+            IQueryable<T> sourceOfEffectiveType = source;
+
+            var effectiveType = GetEffectiveType<T>();
```

Comment:
Can we remove this duplicate code with the ApplyOrder method?

Reply:
Indeed, removed.